### PR TITLE
8325671: Shenandoah: Introduce a ShenandoahGenerationType and templatize certain marking closures with it

### PR DESCRIPTION
**Pull request by ysramakrishna**
/issue JDK-8325671

/summary

In support of eventually supporting a generational version of Shenandoah, we introduce a ShenandoahGenerationType enum which currently has a single NON_GEN value for Shenandoah. The generational extension of Shenandoah will introduce additional enum values when GenShen is integrated. These will be used to specialize certain marking closures via templatization, such that suitable variations of the marking (and in the future other) closures can be used for the generational and non-generational variants of the collector while minimizing interference between the two variants while maximizing code-sharing without affecting correctness or performance.

This ticket introduces the new enum type and templatizes certain existing marking closures.

In effect, this should be semantically a no-op for current (non-generational) Shenandoah and ideally completely performance-neutral (to be established via measurements).

**Testing:**
- [x] jtreg hotspot_gc
- [x] github actions
- [x] codepipeline perf & stress

**Performance:**
- [x] specjbb w/shenandoah
- [x] codepipeline perf

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325671](https://bugs.openjdk.org/browse/JDK-8325671): Shenandoah: Introduce a ShenandoahGenerationType and templatize certain marking closures with it (**Sub-task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [bce39e40](https://git.openjdk.org/jdk/pull/17815/files/bce39e40aaaf437cfdc9d39c8a6d7c3dd369f30f)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17815/head:pull/17815` \
`$ git checkout pull/17815`

Update a local copy of the PR: \
`$ git checkout pull/17815` \
`$ git pull https://git.openjdk.org/jdk.git pull/17815/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17815`

View PR using the GUI difftool: \
`$ git pr show -t 17815`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17815.diff">https://git.openjdk.org/jdk/pull/17815.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17815#issuecomment-1939669051)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.hpp:36]: TBD: Needs #include of ShenandoahGenerationType.

Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.hpp:36]: done.

**Comment by ysramakrishna** [comment]
Did some spot checks and overll sanity checks to compare the libjvm.so contents before and after the change, and don't see any great difference (although the libjvm.so size did increase a tad). Here's one example of the new tempaltized vs old non-templatized versions (second column is the size from nm):

**Comment by ysramakrishna** [comment]
Difference in (release) libjvm.so sizes (second column is size, a difference of ~2.4KB, less than 0.01%):
```                                                                        
4095288896 26968936 generation_type.libjvm.so
2020291223 26966488 master.libjvm.so
```

**Review by kdnilsen — APPROVED** [review]
Thanks for these refinements.  Look good to me.

**Comment by ysramakrishna** [comment]
Can I please get a review from @rkennke , @shipilev , or other Reviewer familiar with Shenandoah. Thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahMark.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahMark.cpp:92]: I think `/* ignored */` is not needed here.

Also, the argument list line is probably too long. AFAICS, we put the arguments we curry from method arguments to template arguments at very end. See how the method arguments are entering the template arguments in the same order:

```
template<bool CANCELLABLE, StringDedupMode STRING_DEDUP>
void ShenandoahMark::mark_loop(uint worker_id, TaskTerminator* terminator, ShenandoahReferenceProcessor *rp,
                               ShenandoahGenerationType generation, StringDedup::Requests* const req) ... 

void ShenandoahMark::mark_loop(uint worker_id, TaskTerminator* terminator, ShenandoahReferenceProcessor *rp,
                               bool cancellable, StringDedupMode dedup_mode, ShenandoahGenerationType generation, StringDedup::Requests* const req ...
```

**Review by shipilev — COMMENTED** [review]
The change looks okay, but I do not see a good reason to introduce this into Shenandoah before Generational mode arrives. It is just extra (mostly dead) code to maintain, and it effectively hides what changes Generational needs to do in shared Shenandoah code. Maybe we should consider this as part of Generational integration PR?

**Comment by ysramakrishna** [comment]
> The change looks okay, but I do not see a good reason to introduce this into Shenandoah before Generational mode arrives. It is just extra (mostly dead) code to maintain, and it effectively hides what changes Generational needs to do in shared Shenandoah code. Maybe we should consider this as part of Generational integration PR?

Indeed the idea was to introduce this earlier that any Generational-specific code as a decomposition and refactoring of the work needed for generational mode, as I describe in the summary above, such that its effect was independently shown to be minimal & harmless, and making subsequent integration of Generational mode much smaller, simpler, and easier to review on its own.

File: src/hotspot/share/gc/shenandoah/shenandoahMark.cpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahMark.cpp:92]: I'll fix the order so it conforms to that convention; thanks for pointing it out!

Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahMark.cpp]: Adjusted order of parms to be consistent with order in templated functions, and fixed long lines.

**Comment by ysramakrishna** [comment]
> Maybe as a compromise we can review this PR now but only ship it with the rest of Generational Shenandoah, when it arrives? Shouldn't be hard to maintain the PR until then, we don't currently expect many changes in the same code. WDYT?

It'd be nicer to be able to ship this so:
1. it gets regular testing on its own
2. it can be shared with subsequent changes in preparation of Generational, which you may have seen more of recently
3. it has the advantage of digestible small incremental chunks of change (pun not intended) that can be easily and independently reviewed on their own without having to produce a series of dependent PRs and maintaining them on a private branch as subsequent work proceeds

Meanwhile, I'll discuss other potential alternatives with the team and think more about the approach I've taken in this regard.

**Comment by ysramakrishna** [comment]
Followed up offline w/Roman, and will check these in later today after a final round of merge-&-test.

Thanks everyone for your reviews and input!

**Pull request closed without merging**
